### DISCUSSION TITLE
improvement idea for `app.openDialog()`

### DISCUSSION THREAD

**Discussion by MiltonBalaOfficial** [xournalpp/xournalpp#6131] [Ideas]
Shifting from `app.msgbox()` to `app.openDialog(message, options, cb, error)` is a fantastic improvement. It has opened up new possibilities for the development of Xournal++ plugins, allowing developers to create more interactive and user-friendly interfaces.

I’ve been brainstorming some ideas for features that could further expand the potential of `app.openDialog`. These are just suggestions, not feature requests, but I wanted to share them with the developers. If any of these ideas resonate and are feasible, they could lead to exciting new use cases.

#### Suggested Features:
1. **Support for Icons Alongside Textual Buttons**  
   It would be amazing to include icons next to the textual buttons in dialogs. For instance, shape icons could be incorporated into plugins like the `ShapeLibrary`, making the buttons more visually intuitive and enhancing usability.

2. **Input Text Field Integration**  
   Adding an input text field directly within the dialog would be a game-changer. Currently, implementing this requires external libraries like `lua-lgi`, which can be a bit complex. If this feature were natively supported, it could enable the creation of richer, standalone plugins with better user interaction capabilities.

---

These are just ideas at this stage, but I believe they have the potential to significantly enhance the flexibility and usability of `app.openDialog`. Thank you for continuing to make Xournal++ more powerful and developer-friendly!

**Comment by rolandlo** [comment]
I have been thinking along these lines, too. The question is how far can we go without adding too much complexity and maintenance cost in the long run. Another idea would be to allow specifying an .ui file and to create UI elements from that file. The question is however how to interact with the thus created UI elements. It's probably too difficult.

**Reply by rolandlo** [reply]
Actually the easier route might be to make `lua-lgi` (and dependencies) part of the Windows installation and of the MacOS .app bundle. We already ship `lua-lgi` with the snap package and with the flatpak on Linux. I will investigate how hard that is and how much storage space it takes.

**Reply by MiltonBalaOfficial** [reply]
>I will investigate

That's great!

**Reply by rolandlo** [reply]
Turns out it wasn't difficult to bundle `lua-lgi` on Windows and also doesn't take much storage space (roughly 4 MB when installed). See #6158.
